blueocean-events/src/main/java/io/jenkins/blueocean/events/PipelineEventListener.java

```diff
@@ -45,135 +44,146 @@
  * This may be useful (or may not be needed) for live-ness of the pipeline results screen.
  */
 @Extension
-public class PipelineEventListener extends RunListener<Run<?,?>> {
+public class PipelineEventListener implements GraphListener {
 
     private static final Logger LOGGER = Logger.getLogger(PipelineEventListener.class.getName());
 
-    private static class StageEventPublisher implements GraphListener {
+    private final Map<FlowExecution, String> currentStageName = new WeakHashMap<>();
+    private final Map<FlowExecution, String> currentStageId = new WeakHashMap<>();
 
-        private final Run run;
-        private final PubsubBus pubSubBus;
-        private String currentStageName;
-        private String currentStageId;
+    @Override
+    public void onNewHead(FlowNode flowNode) {
+        // test whether we have a stage node
+        if (PipelineNodeUtil.isStage(flowNode)) {
```

Comment:
[for reviewers](https://github.com/jenkinsci/blueocean-plugin/pull/1228/files?w=1)